core/src/mender-client.c

```diff
@@ -96,8 +96,8 @@ static const struct mender_update_state_transition_s update_state_transitions[N_
     /* MENDER_UPDATE_STATE_COMMIT                 */ { MENDER_UPDATE_STATE_CLEANUP, MENDER_UPDATE_STATE_ROLLBACK },
     /* MENDER_UPDATE_STATE_CLEANUP                */ { MENDER_UPDATE_STATE_END, MENDER_UPDATE_STATE_END },
     /* MENDER_UPDATE_STATE_ROLLBACK               */ { MENDER_UPDATE_STATE_ROLLBACK_REBOOT, MENDER_UPDATE_STATE_FAILURE },
-    /* MENDER_UPDATE_STATE_ROLLBACK_REBOOT        */ { MENDER_UPDATE_STATE_ROLLBACK_VERIFY_REBOOT, MENDER_UPDATE_STATE_FAILURE },
-    /* MENDER_UPDATE_STATE_ROLLBACK_VERIFY_REBOOT */ { MENDER_UPDATE_STATE_FAILURE, MENDER_UPDATE_STATE_FAILURE },
+    /* MENDER_UPDATE_STATE_ROLLBACK_REBOOT        */ { MENDER_UPDATE_STATE_ROLLBACK_VERIFY_REBOOT, MENDER_UPDATE_STATE_ROLLBACK_VERIFY_REBOOT },
+    /* MENDER_UPDATE_STATE_ROLLBACK_VERIFY_REBOOT */ { MENDER_UPDATE_STATE_FAILURE, MENDER_UPDATE_STATE_ROLLBACK_VERIFY_REBOOT },
```

Comment:
So when a rollback reboot fails, we enter rollback reboot verification which is basically pre-determined to fail because the rollback reboot itself failed. I don't see how that's supposed to work.

Reply:
In the c++ client there is no failure state for rollback reboot, it goes to verification or state loop detection. I guess the idea is that we should always verify the reboot, according to this: https://github.com/mendersoftware/mender/commit/7fb71aab28edf3eed7799ddaca1d01ecd2cf10be

Reply:
I don't care about the C++ client too much, TBH. This is a different world. In the MCU client, a reboot can mean merely restarting a thread or whatever else. So let's focus on the logic here. And I still see a logical error in the flow described above.

Reply:
Ah yes, verify reboot goes to verify reboot - I didn't catch that.  But does it not make sense that we always verify the reboot? Rollback reboot goes to verify rollback reboot - which verifies if it has been successful or not.
If it's not, it retries the rollback reboot - and this will continue until the verification is successful, or 
(when it's implemented) an infinite loop is detected.

Reply:
> Ah yes, verify reboot goes to verify reboot - I didn't catch that.

Yes, that's a short loop. But also a failure in reboot goes to verify reboot -- why verify something that we already know failed?

Reply:
I guess my only motivation is to align it with the c++ client - i.e. follow the flow as closely as possible. What do you think @lluiscampos

Reply:
> I don't care about the C++ client too much

As much as possible, we need to align the behavior of the two clients. This is specially true for the state machine and the Update Modules API. We want is that a user familiar with Mender as a product, that has written maybe an Update Module at some point, can get started with Mender MCU faster and know what to expect.

Keep in mind also that our long term strategy is to merge the two state machines.

So yes, my vote on this one is to align with C++ client: ignore errors from the "do reboot" callback, and call "check reboot" callback.

Reply:
Just to make it clear -- the current implementation will loop between a failing rollback reboot and a failing rollback reboot verification. With no sleep, nothing. When we add some counter/loop detection, it will just quickly exhaust the counter by calling the same callback(s) repeatedly.

Reply:
> but I'm not going to change the C++ client

Me neither.

> the current implementation will loop between a failing rollback reboot and a failing rollback reboot verification

That was an acknowledge typo, fixed now AFAICS.

(this also confused me in the first review, when I then suggested adding an sleep)

Reply:
No, it's still there:
![image](https://github.com/user-attachments/assets/3e0aa644-b63d-4b25-bbe2-9852d90333cb)

A failure in rollback reboot goes to verify rollback reboot which, because the reboot failed, is doomed to fail and goes back to rollback reboot. Or what am I missing here?

Reply:
You are right, and you are not missing anything. Rollback reboot will go to verify rollback reboot regardless. As I said, I just did this because it's how it is done the in the c++ client. It removed the failure state [here](https://github.com/mendersoftware/mender/commit/7fb71aab28edf3eed7799ddaca1d01ecd2cf10be). Seems like it's done this way to align it with the go client. I can agree that the behavior seems strange, but we want to align the state machines, right?

Reply:
>  is doomed to fail and goes back to rollback reboot. Or what am I missing here?

reboot -> verify -> reboot ==> That is correct. Eventually it will break out if we cannot reboot.

Reply:
> Eventually it will break out if we cannot reboot.

That "eventually" will be after a sequence of calls of the two callbacks in a fast sequence with nothing in between them. Except for wearing the flash by writing the new state when switching between them.

Reply:
> I can agree that the behavior seems strange, but we want to align the state machines, right?

Well, yes. But in the C++ client, the reboot, IIRC, is a code in the client itself. And it is a full OS reboot which is very unlikely to fail. Here we don't have a clue what the reboot callback does and how likely it is to fail.

Reply:
Go ahead and merge this if you feel like it implements the correct behavior. There's no point in arguing whether it is or isn't correct. I just wanted to make sure you guys realize what the current implementation will do in case of a failing reboot (callback).

Reply:
I don't get it, @vpodzime. What is a "failed reboot" to you? A "no reboot"?

I think the whole reasoning is that the call might fail but the device is going to reboot anyway, so in my scenario there is no "fast sequence" between these two states.

Reply:
Rollback reboot is a callback, right? And it might fail and perform no reboot. If that happens, we call the verify rollback reboot callback which should fail because there was no (rollback) reboot. And then we call the rollback reboot callback again while nothing has most likely changed since the last attempt unless the scheduler took out the CPU from us and let something else run. So it fails again. And then we call the verify rollback reboot callback which is should again fail (still no reboot). And this repeats basically in a spinlock until some counter or something is exhausted. In the the mean time switching between the states there and back wears flash.

Reply:
If it performed no reboot, the flow that you describe is correct.

The logic is based on the principle that reboot might return failure but still actually reboot, where then such a fast loop won't occur. Granted, this is based on Linux `reboot` command behavior and not applicable on a callback design (on the same binary!), were even an async reboot trigger of some sort should correctly return OK/FAIL.

Let's also look at this from higher level,

We are considering the rollback reboot case, no the regular reboot case where there shall not be retries that will end up in this infinite loop. The rollback reboot state is in the failure path, so it cannot be considered the "normal flow" and it will happen rarely. If a device is not able to rollback reboot, it is likely busted already anyway so wearing flash with ~10 more writes won't really make a difference I believe.

All in all I still think that keeping the two state machines aligned is the requirement that should prevail in this discussion.

Reply:
Cool. As I wrote above, I'm not going to push you on that. I just wanted to make sure we acknowledged the "spin lock" behavior.